### ISSUE TITLE
print_schema.rs throws error for MySQL when enum present in schema

### ISSUE THREAD

**Issue by adwhit** [diesel-rs/diesel#3386]
### Versions

- **Rust:**: 1.64
- **Diesel:** 2.0.1
- **Database:**: MySQL 8
- **Operating System**: MacOS

### Feature Flags

- **diesel:** postgres,mysql,sqlite

## Problem Description

Trying to test my crate https://github.com/adwhit/diesel-derive-enum.

I have a very simple schema to run against mysql (running in a docker container).

```sql
CREATE TABLE simple (
  id SERIAL PRIMARY KEY,
  some_value ENUM('foo', 'bar', 'baz_quxx') NOT NULL
);
```
This is valid mysql, see https://dev.mysql.com/doc/refman/8.0/en/enum.html.

When I execute `diesel-cli` I get an error:

`$ diesel --database-url mysql://root:mysql@127.0.0.1:3306/db migration run`
```
Encountered unknown type for Mysql: enum
thread 'main' panicked at 'internal error: entered unreachable code: Mysql only supports a closed set of types.
                         If you ever see this error message please open an issue at https://github.com/diesel-rs/diesel containing a dump of your schema definition.', diesel_cli-2.0.1/src/print_schema.rs:313:17
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

The same problem came up here: https://github.com/diesel-rs/diesel/discussions/3378 -  unresolved.

## Checklist

- [x] I have already looked over the [issue tracker](https://github.com/diesel-rs/diesel/issues)  and the [disussion forum](https://github.com/diesel-rs/diesel/discussions)  for similar possible closed issues.
 
- [x] This issue can be reproduced on Rust's stable channel. (Your issue will be
  closed if this is not the case)
- [x] This issue can be reproduced without requiring a third party crate

**Comment by weiznich** [comment]
Thanks for opening this issue. This is expected behavior as we do not have any support for the mysql enum type in diesel yet. At least for me its not clear how exactly a type safe abstraction for such types would look like as there is one shared underlying type (`enum`) that is used for fundamentally different types, for example if you define different enum columns (with different possible values) on the same table. We would need a design for handling these situations first. I'm open for suggestions (+ implementations here), but that's likely nothing I personally will work on for the next few releases.

What definitively should be done is improving the error message, so that it states that `enum` is currently unsupported.

**Comment by weiznich** [comment]
I had some time to think a bit more about this. As already written: We want each `enum` field to be it's own type. The correct way to handle this seems to generate a custom type on the fly every time we encounter an `enum` type for the mysql backend. We already do something similar for postgresql, altough it's easier there as the types have names that are defined on the database side. For mysql we would need to generate that name somehow. Likely we can just use the table + column name as identifier as these should be unique. 

The infrastructure for this is mainly already in place. The panic is triggered here: https://github.com/diesel-rs/diesel/blob/master/diesel_cli/src/print_schema.rs#L333

That's the place where we would need to check whether the type name is `enum` and then generate a corresponding custom type for that field. That's basically that what's done for postgres here: https://github.com/diesel-rs/diesel/blob/master/diesel_cli/src/print_schema.rs#L277-L311 
We would need to generate the type name based on the table + column name, so these information needs to be included in the `CustomTypeList` struct. Additionally this import code needs to be adjusted to generate imports for the  generated type: https://github.com/diesel-rs/diesel/blob/master/diesel_cli/src/print_schema.rs#L446

Finally we need to generate the correct type name for the column type here: https://github.com/diesel-rs/diesel/blob/master/diesel_cli/src/print_schema.rs#L564

I'm happy to review a PR implementing this. Feel free to reach out for questions if something is unclear.

**Comment by Abhicodes-crypto** [comment]
Hi guys can I work on this ?

**Comment by weiznich** [comment]
@Abhicodes-crypto Yes, contributions here are welcome and I'm not aware that anyone is working on this yet. Feel free to reach out if anything unclear or if you need help with certain parts of the code-base.

**Comment by Abhicodes-crypto** [comment]
Cool. I am working on this.

**Comment by ParadoxTR** [comment]
Is there any update about this?

**Comment by weiznich** [comment]
@ParadoxTR We are waiting for contributions here. Feel free to submit a PR, instructions on what needs to be changed can be found above.
Otherwise its strongly discouraged to ask for updates, so please do not do that again.

**Comment by peter-todorov** [comment]
I've also started working on this and I've implemented everything described above, but I'm not sure if it's enough. posgresql has it's own derive "#[diesel(postgres_type(name=" which correctly handles code generation. While  implementing what's described above leaves me with bare structs for every mysql enum and set. @weiznich can you please advise me on how should I proceed? Do we need derives for mysql as well? Or are bare structs for each enum and set field enough.  

I'm totally new to Diesel btw, I haven't written a single query yet.

**Comment by weiznich** [comment]
Thanks for working on this issue. The mysql equivalent of [`#[diesel(postgres_type(…))]` is `#[diesel(mysql_type(name = "…")]`](https://docs.diesel.rs/master/diesel_derives/derive.SqlType.html). The provided name needs to be one of the variants from the [`MysqlType`](https://docs.diesel.rs/master/diesel/mysql/enum.MysqlType.html#variant.Enum) enum. In this specific case it always needs to be `Enum`.

**Comment by weiznich** [comment]
That's correct. For support for mysql set we would need to use [`MysqlType::Set`](https://docs.diesel.rs/master/diesel/mysql/enum.MysqlType.html#variant.Set)

**Comment by tedfordgif** [comment]
According to [a comment on #3523](https://github.com/diesel-rs/diesel/pull/3523#issue-1595974972), this should be open, right @Ten0?

**Comment by weiznich** [comment]
Fixed with #3530
